beacon-chain/sync/pending_blobs_queue.go

```diff
@@ -0,0 +1,147 @@
+func (s *Service) processPendingBlobs() {
+	s.waitForChainStart()
+
+	eventFeed := make(chan *feed.Event, 1)
+	sub := s.cfg.stateNotifier.StateFeed().Subscribe(eventFeed)
+	defer sub.Unsubscribe()
+
+	// Initialize the cleanup ticker at 5s and 10s mark. The node is less busy that time.
+	cleanupTicker := slots.NewSlotTickerWithIntervals(s.cfg.chain.GenesisTime(), []time.Duration{5, 10})
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-sub.Err():
+			return
+		case e := <-eventFeed:
+			s.handleEvent(s.ctx, e)
+		case <-cleanupTicker.C():
+			if s.pendingBlobSidecars == nil {
+				return
+			}
+			s.pendingBlobSidecars.cleanup()
+		}
+	}
+}
+
+// handleEvent processes incoming events.
+func (s *Service) handleEvent(ctx context.Context, e *feed.Event) {
+	if e.Type == statefeed.BlockProcessed {
+		s.handleNewBlockEvent(ctx, e)
+	}
+}
+
+// handleNewBlockEvent processes blobs when a parent block is processed.
+func (s *Service) handleNewBlockEvent(ctx context.Context, e *feed.Event) {
+	data, ok := e.Data.(*statefeed.BlockProcessedData)
+	if !ok {
+		return
+	}
+	if data == nil || data.SignedBlock.IsNil() {
+		return
+	}
+	s.processBlobsFromSidecars(ctx, data.SignedBlock.Block().ParentRoot())
+}
+
+// processBlobsFromSidecars processes blobs for a given parent root.
+func (s *Service) processBlobsFromSidecars(ctx context.Context, parentRoot [32]byte) {
+	blobs := s.pendingBlobSidecars.pop(parentRoot)
+	for _, blob := range blobs {
+		if err := s.receiveBlob(ctx, blob); err != nil {
+			log.WithError(err).Error("Failed to validate blob in pending queue")
+		}
+	}
+}
+
+// receiveBlob validates and processes a blob.
+func (s *Service) receiveBlob(ctx context.Context, blob *eth.SignedBlobSidecar) error {
+	result, err := s.validateBlobPostSeenParent(ctx, blob)
+	if err != nil {
+		return err
+	}
+	if result != pubsub.ValidationAccept {
+		return fmt.Errorf("unexpected pubsub result: %d", result)
+	}
+	return s.cfg.chain.ReceiveBlob(ctx, blob.Message)
+}
+
+// blobWithExpiration holds blobs with an expiration time.
+type blobWithExpiration struct {
+	blob      []*eth.SignedBlobSidecar
+	expiresAt time.Time
+}
+
+// pendingBlobSidecars holds pending blobs with expiration.
+type pendingBlobSidecars struct {
+	sync.RWMutex
+	blobSidecars map[[32]byte]*blobWithExpiration // Key is the block root.
+}
+
+// newPendingBlobSidecars initializes a new cache of pending blobs.
+func newPendingBlobSidecars() *pendingBlobSidecars {
+	return &pendingBlobSidecars{
+		blobSidecars: make(map[[32]byte]*blobWithExpiration),
+	}
+}
+
+// add adds a new blob to the cache.
+func (p *pendingBlobSidecars) add(blob *eth.SignedBlobSidecar) {
+	p.Lock()
+	defer p.Unlock()
+	parentRoot := bytesutil.ToBytes32(blob.Message.BlockParentRoot)
+	expirationTime := time.Now().Add(time.Duration(2*params.BeaconConfig().SecondsPerSlot) * time.Second)
+
+	if existing, exists := p.blobSidecars[parentRoot]; exists {
+		existing.blob = append(existing.blob, blob)
```

Comment:
I guess this relies on libp2p to assure that we do not see duplicates? Otherwise I would suggest a map keyed by blob index or commitment.

Reply:
I dont like the DOS vector, I'll add a check for index